### DISCUSSION TITLE
[MTR05] week1 作業挑戰題，如果使用者傳入的不是數字

### DISCUSSION THREAD

**Discussion by Jane0901** [Lidemy/forum#71] [程式導師實驗計畫] [answered]
成功完成**傳入一個數字 n，然後會產生 1~n 個檔案**的挑戰題
```
#!/bin/bash

for ((i=1 ; i<=$1 ; i++))
do
  touch ${i}.js;
done
```
而助教有給「如果使用者傳入的不是數字會發生什麼事呢？」的延伸方向。
我原本以為會出現 error，結果實際測試後發現沒有產生新檔案也沒有跑出 error，然後把 $1 換成隨便的詞也確實是同樣結果，但並沒有很明白其中的道理，所以想請問背後的原因是什麼呢？
麻煩大家解惑了，謝謝:pray:

**Comment by aszx87410** [comment, marked as answer]
我猜是因為型態轉換相關的東西，導致 i<=$1 是 false，所以就不會執行迴圈但也不會壞掉
但這要詳細查一下才知道

可以 tag 當初問你這問題的助教XD

**Reply by SimonOneNineEight** [reply]
是我問的！但是我原本也預期會噴錯XD
但是結果不會所以我就請他上來問看看有沒有更瞭解的人

**Reply by aszx87410** [reply]
只好召喚一下感覺助教中對 shell script 比較熟的 @JAS0NHUANG 出來看看XD

**Reply by JAS0NHUANG** [reply]
老師有時候「感覺」還是會出錯 XD。
我對 shell script 也沒有比較熟，所以就查了點資料，然後自已做了一點實驗。
大概可以分成兩個部分：
#### 1. Arithmetic expressions 裡的「數字」與「非數字」
`(( ))` 在 shell script 裡面是用來處理 Arithmetic expressions，也就是與數字有關的「計算」、「賦值」、「比較」，如果傳入的值不是數字，那它就會把它看成是 0 然後進行運算或比較，其實就有點像是老師在上面提到的型態轉換，但是它的行為是直接把「非數字」轉換成數字 0。
```bash
myName="Jas0n"
echo $(( myName + 99 )) #印出 99
echo $(( 1 + 99 )) #印出 100
```
用在比較或是 `for` 迴圈裡行為也差不多，在 `for` 迴圈裡其實就是用分號分開的三個  Arithmetic expression，第一個試定初始值、第二個比較、第三個加加並重新賦值，如果你傳入的是「非數字」的值（如上面的 `myName`），在第二個 expression 就會被當成 0 去比較，所以 `for (( i = 1; i <= myName ; i++))` 是不會出錯的，只是不會進入迴圈（這裡你可以試試看寫成 `for (( i = 0; i <= $1; i++))`，如果給的值不是數字的話，是會生成一個 `0.js` 檔案的）。

#### 2. 「計算」、「賦值」、「比較」與 command line 參數
關於 Arithmetic expressions 的「計算」、「賦值」與「比較」行為也不一樣，單純「計算」的話，不給 command line 參數也不會出錯，所以 `myScript.sh` 裡面寫 `echo $(( $1 + 99 ))` 然後直接執行 `./myScript.sh`，就算不給參數還是可以印出答案不會出錯。
但是進行「比較」、「賦值」時，不給參數就會噴錯，所以在 `for` 迴圈裡如果沒有給 command line 參數就會出現 `syntax error`，只要給了參數，不論參數的型態是什麼都沒關係了！

至於它們真正的底層原理我也還沒有很清楚……只知道它的行為大概是這樣囉！

雖然老師「感覺」有些失誤，但是都被召喚了，只好硬著頭皮入坑 XD，我印象中有幾個同學是寫後端、做 DevOps 的？應該比我更清楚，真的有講錯或是解釋不夠仔細的地方再拜託指正、補充了。

參考資料：
[Bash Reference Manua - Shell Arithmetic](https://www.gnu.org/software/bash/manual/html_node/Shell-Arithmetic.html)
[Bash Hackers Wiki - Arithmetic expressions](https://wiki.bash-hackers.org/syntax/arith_expr)

**Reply by Jane0901** [reply]
看完助教仔細地講解後有通了很多的感覺！
可以瞭解這些行為已經很滿足了哈哈
感謝老師、助教們的回覆:bowing_woman:
